pcap-dpdk.c

```diff
@@ -0,0 +1,811 @@
+			if (nb_rx){ // got packets within timeout_ms
+				break;	
+			}else{ // no packet arrives at this round.
+				if (p->break_loop){
+					break;
+				}
+				// sleep for a very short while.
+				// block sleep is the only choice, since usleep() will impact performance dramatically.
+				rte_delay_us_block(DPDK_DEF_MIN_SLEEP_MS*1000);
+				sleep_ms += DPDK_DEF_MIN_SLEEP_MS;
+			}
+		}
+	}
+	return nb_rx;
+}
+
+static int pcap_dpdk_dispatch(pcap_t *p, int max_cnt, pcap_handler cb, u_char *cb_arg)
+{
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	int burst_cnt = 0;
+	int nb_rx = 0;
+	struct rte_mbuf *pkts_burst[MAX_PKT_BURST];
+	struct rte_mbuf *m;
+	struct pcap_pkthdr pcap_header;
+	uint16_t portid = pd->portid;
+	// In DPDK, pkt_len is sum of lengths for all segments. And data_len is for one segment
+	uint32_t pkt_len = 0;
+	int caplen = 0;
+	u_char *bp = NULL;
+	int i=0;
+	unsigned int gather_len =0;
+	int pkt_cnt = 0;
+	int is_accepted=0;
+	u_char *large_buffer=NULL;
+	int timeout_ms = p->opt.timeout;
+
+	if ( !PACKET_COUNT_IS_UNLIMITED(max_cnt) && max_cnt < MAX_PKT_BURST){
+		burst_cnt = max_cnt;
+	}else{
+		burst_cnt = MAX_PKT_BURST;
+	}
+
+	while( PACKET_COUNT_IS_UNLIMITED(max_cnt) || pkt_cnt < max_cnt){
+		if (p->break_loop){
+			p->break_loop = 0;
+			return PCAP_ERROR_BREAK;
+		}
+		// read once in non-blocking mode, or try many times waiting for timeout_ms.
+		// if timeout_ms == 0, it will be blocked until one packet arrives or break_loop is setted.
+		nb_rx = dpdk_read_with_timeout(p, portid, 0, pkts_burst, burst_cnt);
+		if (nb_rx == 0){ 
+			if (pd->nonblock){
+				RTE_LOG(DEBUG, USER1, "dpdk: no packets available in non-blocking mode.\n");
+			}else{
+				if (p->break_loop){
+					RTE_LOG(DEBUG, USER1, "dpdk: no packets available and break_loop is setted in blocking mode.\n");
+					p->break_loop = 0;
+					return PCAP_ERROR_BREAK;
+
+				}
+				RTE_LOG(DEBUG, USER1, "dpdk: no packets available for timeout %d ms in blocking mode.\n", timeout_ms);
+			}
+			// break if dpdk reads 0 packet, no matter in blocking(timeout) or non-blocking mode.
+			break;	
+		}
+		pkt_cnt += nb_rx;
+		for ( i = 0; i < nb_rx; i++) {
+			m = pkts_burst[i];
+			calculate_timestamp(&(pd->ts_helper),&(pcap_header.ts));
+			pkt_len = rte_pktmbuf_pkt_len(m);
+			// caplen = min(pkt_len, p->snapshot);
+			// caplen will not be changed, no matter how long the rte_pktmbuf
+			caplen = pkt_len < p->snapshot ? pkt_len: p->snapshot; 
+			pcap_header.caplen = caplen;
+			pcap_header.len = pkt_len; 
+			// volatile prefetch
+			rte_prefetch0(rte_pktmbuf_mtod(m, void *));
+			bp = NULL;
+			if (m->nb_segs == 1)
+			{
+				bp = rte_pktmbuf_mtod(m, u_char *);
+			}else{
+				// use fast buffer pcap_tmp_buf if pkt_len is small, no need to call malloc and free
+				if ( pkt_len <= ETHER_MAX_JUMBO_FRAME_LEN)
+				{
+					gather_len = dpdk_gather_data(pd->pcap_tmp_buf, RTE_ETH_PCAP_SNAPLEN, m);
```

Comment:
Discaimer: I'm do not know much about dpdk, my assumtion making this comment is that the packet(s) are in linked mbuffs and a packet may span 2 consecutive bugffers.

I think "zero-copy" would be much faster e.g pass the mbuff to the caller and let the writing code handle the case of a packet continuing in the next buffer. Reading about dpdk I got the impression that it was possible to create a header section in the buffer before the actual packet. If that is true it would be very efficent to leave space for a pcap/pcap-ng packet header in the buffer in that way the whol buffer co0uld be written to file in one go.

Reply:
This link may be interesting too https://www.napatech.com/dpdk-packet-capture-pdump/

Reply:
That article claims

> Basically the pcap_dump() function does a fwrite(pcap header of 16Bytes) followed by a fwrite(packet payload) and these two fwrite(s) cause two kernel traps, which are expensive and should be minimized.

which is an over-simplification.  `fwrite()` does not necessarily cause a kernel trap; with the standard I/O library, a kernel trap occurs only if it does a `write()` (on UN\*X) or an `NtWriteFile()` (on Windows; that will come from a Win32 `WriteFile()`, which would be done by an `_write()`), which will happen only in an `fwrite()` that requires that the buffer be emptied to make enough room to copy the data.

The typical buffer size for the standard I/O library is, these days, probably at least 4KiB, which is sufficient for at least one pcap record for a maximum-size Ethernet packet.  libpcap could increase that buffer size by making a `setvbuf()` call, which would reduce the number of system calls done.  I've sent the author fo that document a e-mail pointing that out and asking how big the buffer in question was.

Reply:
> I think "zero-copy" would be much faster e.g pass the mbuff to the caller and let the writing code handle the case of a packet continuing in the next buffer.

It's more like "...to the *callee*"; libpcap's core captured-packet-processing loops pass the packets to a callback routine.  Those routines are expecting the packet data to be in a contiguous array of size equal to the captured length, so if the packet is split between mbufs, it would have to be coalesced into a single buffer.

If the goal is to have a special fast path for dumping, that would have to be done through additional pcap APIs, e.g. a `pcap_dispatch_and_dump()` API that takes a `pcap_t` and a `pcap_dumper_t` as arguments and dumps each packet to the `pcap_dumper_t`.  That would possibly involve running a different captured-packet-processing loop for capture modules where packet coalescing is done (for other modules, `pcap_dispatch_and_dump()` would just be something such as

    int
    default_pcap_dispatch_and_dump_handler(pcap_t *p, int cnt, pcap_dumper_t *pdmp)
    {
       return pcap_dispatch(p, cnt, pcap_dump, pdmp);
    }

so that we don't need two loops for *all* pcap-XXX.c files).